### ISSUE TITLE
Since 1.6.0 TypeVar with bound types fails to match with Callable of the same bound types

### ISSUE THREAD

**Issue by anselor** [python/mypy#16374]
**Bug Report**

Did not occur prior to 1.6.0.

https://github.com/python-cmd2/cmd2/actions/runs/6699406223/job/18203570049
Got the following error with mypy:
```
cmd2/cmd2.py: note: In member "_build_parser" of class "Cmd":
cmd2/cmd2.py:699:46: error: Argument 1 has incompatible type "CommandParent"; expected "Union[Type[Cmd], Type[CommandSet]]"  [arg-type]
Found 1 error in 1 file (checked 18 source files)
```
where CommandParent is a TypeVar declared as:
```python
CommandParent = TypeVar('CommandParent', bound=Union['cmd2.Cmd', CommandSet])
```

**To Reproduce**

https://github.com/python-cmd2/cmd2/blob/8d88c357ca0764b114c68d8175d2cd33146b83d7/cmd2/cmd2.py#L694

**Expected Behavior**

Since CommandParent is bound to the expected type, this shouldn't be an error

**Actual Behavior**

I get the error above.

**Your Environment**

- Mypy version used: 1.6.1
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.8, 3.9,3.10,3.11

**Comment by erictraut** [comment]
Could you provide a minimal, self-contained code sample that exhibits the problem?
